File: src/main/java/me/lokka30/levelledmobs/rules/MergeableStringList.java
package me.lokka30.levelledmobs.rules;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedList;
import java.util.List;
import org.jetbrains.annotations.Nullable;


@SuppressWarnings("unused")
public class MergeableStringList {

    MergeableStringList() {
    }

    MergeableStringList(final @Nullable String item) {
        if (item == null) {
            return;
        }

        this.items = new LinkedList<>();
        this.items.add(item);
    }

    MergeableStringList(final @Nullable String item, final boolean doMerge) {
        if (item == null) {
            return;
        }

        this.items = new LinkedList<>();
        this.items.add(item);
        this.doMerge = doMerge;
    }

    public List<String> items;
    boolean doMerge;

    public void setItemFromString(final @Nullable String input) {
        if (input == null) {
            return;
        }

        this.items = new ArrayList<>(1);
        this.items.add(input);
    }

    void setItemFromList(final @Nullable Collection<String> input) {
        if (input == null) {
            return;
        }

        this.items = new LinkedList<>();
        this.items.addAll(input);
    }

    public void mergeFromList(final @Nullable Collection<String> input) {
        if (input == null) {
            return;
        }

        if (this.items == null) {
            this.items = new LinkedList<>();
        }
        this.items.addAll(input);
    }

    public boolean isEmpty() {
        return this.items.isEmpty();
    }

    public boolean isNotEmpty(){
        return !this.isEmpty();
    }

    public String toString() {
        if (items == null || items.isEmpty()) {
            return super.toString();
        }

        final StringBuilder sb = new StringBuilder();
        if (items.size() == 1) {
            sb.append(items.get(0));
        } else {
            sb.append(items);
        }

        if (doMerge) {
            if (!sb.isEmpty()) {
                sb.append(" ");
            }
            sb.append("(merge)");
        }

        return sb.toString();
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/rules/MergeableStringList.kt
package io.github.arcaneplugins.levelledmobs.rules

/**
 * Provides a list of strings that can be optionally
 * merged together
 *
 * @author stumper66
 * @since 3.3.0
 */
class MergeableStringList {
    val items = mutableListOf<String>()
    var doMerge: Boolean = false

    constructor()

    constructor(item: String?){
        if (item != null)
            this.items.add(item)
    }

    constructor(item: String?, doMerge: Boolean){
        if (item == null) return

        items.add(item)
        this.doMerge = doMerge
    }

    fun setItemFromString(input: String?) {
        if (input == null) return

        items.add(input)
    }

    fun setItemFromList(input: Collection<String>?) {
        if (input == null) return

        items.addAll(input)
    }

    fun mergeFromList(input: Collection<String>?) {
        if (input == null) return

        items.addAll(input)
    }

    val isEmpty: Boolean
        get() = items.isEmpty()

    val isNotEmpty: Boolean
        get() = !this.isEmpty

    override fun toString(): String {
        if (items.isEmpty()) return super.toString()

        val sb = StringBuilder()
        if (items.size == 1)
            sb.append(items[0])
        else
            sb.append(items)

        if (doMerge) {
            if (sb.isNotEmpty()) sb.append(" ")

            sb.append("(merge)")
        }

        return sb.toString()
    }
}